# Jugando Atari con Julia y Reinforcement Learning

No nos emocionemos aun, todavia hay un largo camino por recorrer... Entrenar un programa (agente) que aprenda a jugar a nivel humano toma mucha paciencia y horas de entrenamiento; un solo paso equivocado y no lo vamos a lograr. 

Para aprender a resolver problemas complejos, lo primero que tenemos que hacer es comprender las complicaciones detras de cada parte de esta tarea. Para esto, necesitamos comenzar con problemas sencillos y escalarlos poco a poco. Necesitamos comenzar haciendo un pequeño viaje en el tiempo a la historia reciente de la Inteligencia Artificial y el Reinforcement Learning.

## 1. MDPs

Nuestra historia comienza con Richard Bellman (*A Markovian Decission Process* 1957) y Ronald Howard (*Dynamic Programming and Markov Processes* 1960), que extendio la teoria de Bellman.

 <center><table><tr>
  <td><img src="Images/bellman.jpg" alt="Richard Bellman" width="150">
  <figcaption>Fig1.a - El es Bellman.</figcaption></td>
  <td><img src="Images/howard.jpg" alt="Ronald Markov" width="150">
  <figcaption>Fig1.b - El es Howard.</figcaption></td>
  <td><img src="Images/markov.jpg" alt="Andrey Markov" width="150">
  <figcaption>Fig1.c - El es Markov.</figcaption></td>
</table></center>

Bellman tuvo una idea muy simple de como modelar un proceso de aprendizaje. En cada momento del tiempo $t$, un agente debe tomar una decision (accion) $A_t$ dependiendo del estado del ambiente $S_t$; en el momento que el agente toma su decision, interactua con el ambiente y lo transforma, llevando al sistema a un nuevo estado $S_{t+1}$. ¿Como puede saber el agente si actuo correcta o incorrectamente? Para eso necesita una señal de reforzamiento o pago. Por esta razon, suponemos ademas que el agente recibe un pago $R_{t+1}$ en el momento que el sistema se transforma.

 <center>
 <img src="Images/mdp_process.png" alt="MDP Representation" width="400" height="225">
 <figcaption>Fig2. - Este es el ciclo de una tarea de aprendizaje.</figcaption>
</center>

¿Y que tiene que ver con Markov? Bellman es uno de los desarrolladores de la *Programacion Dinamica*, que es una tecnica para resolver problemas matematicos en los que las variables que queremos encontrar dependen del tiempo y la solucion en un momento $t$ se puede expresar en terminos de la solucion al tiempo $t-1$. La idea central del trabajo de Bellman consiste en que para encontrar las decisiones optimas $A_t, A_{t+1},...$ de forma que maximicen el total de pagos recibidos en el tiempo $R_{t+1}, R_{t+2}, ...$ tenemos que suponer que el estado futuro del sistema es independiente del pasado condicionado a su estado actual y de las decisiones del agente. Esta propiedad se conoce como la propiedad de Markov. En terminos matematicos, la implicacion de esta condicion es
$$
P(S_{t+1}, R_{t+1} \mid S_0, A_0, R_0, ..., S_t, A_t, R_t) = P(S_{t+1}, R_{t+1} \mid S_t, A_t).
$$
En otras palabras, las probabilidades de transicion $(S_t,A_t) \to (S_{t+1}, R_{t+1})$ y las decisiones del agente determinan todo el comportamiento del sistema. Un problema reforzamiento con la propiedad de Markov se conoce como un *Markov Decission Process* o MDP.

## Un juego de Atari como un MDP

<center>
 <img src="Images/pong.jpg" alt="Pong" width="150">
 <figcaption>Fig3. - El juego de de Atari Pong.</figcaption>
</center>

Este tutorial es acerca de como resolver juegos de Atari en Julia. Para promover el desarollo de algoritmos de Inteligencia Artificial y Reinforcement Learning, la organizacion OpenAI creo la interfaz [Open AI Gym](https://gym.openai.com/) que facilita el desarollo y evaluacion de algoritmos de aprendizaje. La mala noticia es que solo esta disponible para Python y es casi imposible de instalar en Windows. La buena noticia es que Julia tiene una interfaz para interactuar con Python de forma muy sencilla y no sera una limitacion.

Por supuesto que pueden optar por implementar sus algoritmos directamente en Python y no en Julia. Pero este es un tutorial de Julia, que explota la belleza, facilidad y velocidad de Julia.

Por favor vean la pagina de Open AI Gym directamente para instrucciones de como instalar la libreria `gym` de Python. Para poder usar `gym` en Julia, deberan antes tener la paqueteria de Interfaz con Python llamada `PyCall`. Para instalarla, usen el comando `Pkg.add("PyCall")`.

El siguiente codigo muestra como cargar el juego de Pong y hacer unas cuantas simulaciones.

In [1]:
using PyCall # use Pkg.add("PyCall") to install
@pyimport gym # most have gym installed on your computer's Python distribution
pong = gym.make("Pong-v0") # crea el ambiente de juego Pong en el objeto pong

[2017-05-10 22:24:51,388] Making new env: Pong-v0


PyObject <TimeLimit<AtariEnv<Pong-v0>>>

Un MDP tiene estados y acciones disponibles, para saber cuales son con un objeto del tipo Pong, podemos usar las acciones de la paqueteria de Python gym.

### Conjunto de Estados

In [2]:
pong[:observation_space] # pong.observation_space en Python

PyObject Box(210, 160, 3)

El resultado es que el espacio de estados es una "caja" de dimensiones $210 \times 160 \times 3$. Aunque en los metodos de Reinforcement Learning no necesitaremos saber el significado de los estados,  en este caso sabemos de antemano que va el juego va a regresar los componentes RGB de los pixeles del juego. El juego tiene una resolucion de $210 \times 160$ y por cada uno tiene tres componentes de color. Un componente de color es un valor en el rango $0--255$. Si no supieramos esto, pudiese haber usado las siguientes funciones para determinar la frontera del espacio de observaciones
* `pong[:observation_space][:high]`
* `pong[:observation_space][:low]`

No las incluimos aqui porque el resultado es muy grande.

### Conjunto de Acciones

Determinar el conjunto de acciones es muy similar, para eso podemos usar

In [3]:
pong[:action_space] # pong.action_space en Python

PyObject Discrete(6)

En este caso es discreto (el conjunto de acciones siempre es discreto en estas pruebas). Eso quiere decir que solo puede obtener los valores $0-5$. Si! Empieza en cero, es una de las herencias por la interfaz con Python.

Otra manera de encontrar el numero de acciones es simplemente con el comando

In [4]:
pong[:action_space][:n] # pong.action_space en Python

6

Tanto el espacio de observacion y espacio de accion son objetos de Python que pertenecen a la clase "espacios" y tienen implementado un metodo muy util para tomas muestras aleatorias de estos espacios. Eso sera muy util al principio, para poder demostrar algunas simulaciones.

In [5]:
pong[:action_space][:sample]()

4

Afortunadamente para los metodos de Reinforcement Learning, tampoco necesitamos saber el significado de estas acciones.

### Simulando transiciones del MDP

Ahora lo mas divertido, veamos primero como simular una transicion usando el metodo `step` de los ambientes de `gym`.

In [6]:
initial_state = pong[:reset]() # comienza oficialmente el ambiente Pong poniendo en su estado inicial
action = pong[:action_space][:sample]() # escoje una accion al azar, ojo no depende del estado inicial
    # por ser completamente aleatorio, pero en principio deberia...
new_state, reward, done, info = pong[:step](action)
; # para que no imprima el resultado de la celda

El metodo `step` recibe una accion (en realidad, tambien recibe el estado actual, para ya esta codificado dentro del ambiente) y regresa cuatro cosas
* `new_state`: El nuevo estado del ambiente
* `reward` El pago recibido por la accion decididad
* `done` Una variable booleana que determina si el episodio esta terminado 
* `info` Un diccionario que trae informacion adicional del juego, en teoria nunca deberiamos necesitarla.

La funcion `step` tiene implica las probabilidades de transicion $(S_t, A_t)\to (S_{t+1}, R_{t+1})$ del MDP. No son explicitas porque no regresa toda la distribucion, simplemente un muestreo de esa distribucion. Esta es la dificultad a la que nos enfrentamos.

### Simulando y visualizando todo un episodio 

Veamos ahora como podriamos simular y visualizar un episodio entero. Para visualizar el estado del ambiente necesitaremos el metodo `render`. 

In [7]:
initial_state = pong[:reset]() 
done = false
episode_reward = 0.
while !done # cicla hasta que la funcion step marque el fin del episodio
    pong[:render]() # crea la visualizacion del juego
    action = pong[:action_space][:sample]() 
    state, reward, done, info = pong[:step](action)
    episode_reward += reward
end
println("Episodio terminado con pago total: ", episode_reward)
pong[:render](close = true) # cierra la visualizacion del juego 

Episodio terminado con pago total: -20.0


**Felicidades!!!** Con esta parte del tutorial, ya sabes todo lo que necesitabas de los ambientes de `gym`! 

Desafortunadamente, resolver el juego de Pong no es facil pues el espacio de estados viene en terminos de pixeles y eso significa muchas posibilidades (aun cuando no las veamos todas). De momento, trataremos de resolver un problema mas sencillo, primero deberan hacer el siguiente ejercicio.

### <span style='color: red'> Ejercicio 1 </span>

Antes de poder juegos de Atari, vamos a utilizar el ambiente "MountainCar-v0" que es una muy sencilla y clasica prueba de algoritmos de Inteligencia Artificial. El problema consiste en un coche que tiene que subir una colina pero su motor no tiene la fuerza necesaria. Para esto tiene que ir de lado en lado construyendo inercia.

<center>
 <img src="Images/mountainCar.jpg" alt="Mountain Car" width="300" height="225">
 <figcaption>Fig3. - Problema del coche de la colina.</figcaption>
</center>
 
Introduce el codigo para cargar el ambiente "MountainCar-v0" y realiza la simulacion de 10 episodios tomando decisiones completamente aleatorias con la funcion `sample` del espacion del ambiente. Tu codigo debe imprimir 
1. El pago total acumulado por cada episodio
2. El total de pasos por episodio
3. El tiempo total por episodio, puedes usar las funciones tic() y toq() de julia para esto.



# 2. Control optimo

La regla de toma de decision del agente se conoce como *politica*. La regla no tiene porque ser deterministica, y de hecho, matematicamente conviente pensar esta regla como una distribucion de probabilidad sobre todas las posibles acciones. Para ser mas preciso, una politica es una forma de convertir el estado actual $S_t$ en un accion $A_t$, es decir, es una familia de distribuciones condicionales $\pi(A_t \mid S_t) := P(A_t \mid S_t)$.

El problema del control optimo es el de encontrar una politica $\pi^*$ para un MDP que maximice la ganancia total 
$$
\pi^* = \mathrm{argmax}_\pi E \left[ R_{t+1} + \gamma R_{t+2} + \gamma^2 R_{t+3} \cdots  \gamma^{T-1} R_T\right].
$$
En esta formula, necesitamos una esperanza porque la suma total de ganancias es aleatoria. El simbolo $\gamma$ es una constante $0<\gamma\leq 1$ que representa la preferencia intertemporal. Cuando $\gamma < 1$, se valora mas un retorno en el presente que en el futuro; nosotros supondremos $\gamma = 1$ por simplicidad.

Teoricamente, el trabajo de Bellman y Howard de Programacion Dinamica para MDPs mostro que siempre es posible encontrar la politica optima si se conoce completamente las probabilidades de transicion $P(S_{t+1}, R_{t+1}| S_t, A_t)$ del MDP. En la practica, enfrentamos un problema adicional: normalmente el agente intentando aprender no conoce las probabilidades de transicion del MDP y por lo tanto no puede recurrir a los metodos exactos de la programacion dinamica para encontrar una politica optima; mas aun, aun cuando si pueda conocer estas probabilidades, en ocasiones la cantidad de estados del sistema es tan grande que los metodos analiticos de la programacion dinamica no son suficiente (por ejemplo, se estima que una partida de ajedrez puede tener hasta $10^{45}$ configuraciones distintas, totalmente intratable para los metodos exactos).


## Las Ecuaciones de Bellman

Aunque los metodos de Programacion Dinamica son frecuentemente impracticos para buscar politicas optimas, son la piedra angular para entender las estrategias exitosas de solucion. Si un sistema se encuentra en un estado $S_t$, el valor esperado que recibira el agente al tomar la decision $A_t$ es
$$
Q_\pi(S_t,A_t) := E_\pi\left[ R_{t+1} + \gamma R_{t+2} + \gamma^2 R_{t+3} \cdots  \gamma^{T-1} R_T \mid S_t, A_t \right]
$$
Las ecuaciones de Bellman son ecuaciones recursivas en terminos de la *la funcion de valor estado-accion $Q$*. La propiedad de Markov garantiza que en cualquier momento del tiempo, el futuro del sistema depende solo de su estado actual y de la decision del agente. Como consecuencia de esta "perdida de memoria", despues de una transicion $(S_t, A_t) \to (S_{t+1}, R_{t+1})$ y una proxima accion $A_{t+1}$ del agente, el sistema vuelve a empezar, por lo que se tiene la identidad
$$
Q_\pi(S_t,A_t) = E_\pi\left[R_{t+1} + \gamma Q_\pi(S_{t+1}, A_{t+1})]\right] \quad\quad \text{(Ecuacion de Bellman 1)}
$$
La ecuacion anterior debe una historia:

<blockquote> El agente (con politica) $\pi$ se encuentra en el estado $S_t$ y decide la accion $A_t$. Justo en este momento tiene una promesa de pago futuro de $Q_\pi(S_t, A_t)$. Tras decidir, el sistema transiciona al estado $S_{t+1}$ recibiendo un pago recibe un pago de $R_{t+1}$ por su decision. Ahora tendra que decidir nuevamente una accion $A_{t+1}$ y tendra una promesa de pago futuro de $Q_\pi(S_{t+1}, A_{t+1})$. El pago $R_{t+1}$ que recibio mas la esperanza de pago futuro $Q_\pi(S_{t+1}, A_{t+1})$ deben de ser igual a la promesa de pago $Q_\pi(S_t, A_t)$ que tenia entes de la transicion.
</blockquote>

Esta ecuacion es la primera de las ecuaciones de Bellman y se cumple para todas las politicas $\pi$. La utilidad de esta ecuacion es evaluar la politica actual $\pi$. 

Aunque existen formar de mejorar la politica $\pi$ una vez que se conoce $Q_\pi$, Bellman tambien encontro una ecuacion recursiva que se cumple unicamente por la politica optima $\pi^*$ y que puede ser usada para encontrar la politica optima directamente. La segunda ecuacion o *ecuacion de optimalidad de Bellman* dice

$$
Q_{\pi^*}(S_t,A_t) = E_{\pi^*}\left[ R_{t+1} + \gamma \max_A Q_{\pi^*}(S_{t+1}, A)]\right] \quad\quad \text{(Ecuacion de Bellman 2)}.
$$

El mecanismo detras de esta expresion recursiva descubierta por Bellman se basa en que un individuo racional siempre escoge la accion que maximiza su utilidad esperada, que en terminos de la funcion $Q_{\pi^*}$ implica
$$
\pi^*(S_t) = \mathrm{argmax}_A Q_{\pi^*}(S_t,A).
$$


## Reinforcement Learning

El *Reinforcement Learning* (RL) busca resolver las limitaciones que de la programacion Dinamica tradicional para resolver MDPs. En RL tener un modelo del ambiente (probabilidades de transicion) es optativo y sirve para hacer planeacion y mejorar los algoritmos, pero no es necesario. Las estrategias de Reinforcement Learning evaluan una politica y la mejoran siguiendo estrategias iterativas. 

Podriamos agrupar a los diversos metodos del RL en tres grupos dependiendo de su grado de "supervision":

1. **Metodos de Programacion Dinamica**: Utilizan de forma directa las ecuaciones de Bellman y necesitan de un conocimiento de las probabilidades de transicion $P(S_{t+1},A_{t+1}\mid S_t, A_t)$, de lo contrario, pueden intentar estimarlas estadisticamente a traves de simulacion. Aunque son muy precisos, son altamente demandantes en terminos de memoria y poder computacional, y requieren de un conocimiento de las probabilidades de transicion del ambiente que rara vez se tienen completamente disponibles.
2. **Metodos de Funciones de Valor**: En vez de intentar conocer las probabilidades de transicion, estos metodos buscan estimar la funcion de valor $Q$ de la politica optima usando la ley de los grandes numeros (LGN). Una vez estimada la funcion Q optima, podemos recuperar la politica optima poniendo $\pi^*(S_t) = \mathrm{argmax}_A Q_{\pi^*}(S_t,A)$. Mas adelante veremos en que formas las Ecuaciones de Bellman y la LGN permiten conocer $Q$.
3. **Metodos de Gradientes de Politica**: Este ultimo grupo es el menos supervisado y trata directamente  de encontrar unas distribuciones de probabilidad $\pi(A \ S)$ para cada pareja estado-accion $(S,A)$, sin pasar por la funcion $Q$, y tratando de maximizar la utilidad acumulada $R_{t+1} + \gamma R_{t+2} + \cdots + \gamma^{T-1}$. Es es el metodo buscar patrones en los estados $S_t$ que directamente se correlacionen con mayores pagos.

Ambos enfoque 2 y 3 son muy populares actualmente. La ventaja de los ultimos es que tienen menos supuestos al ser menos estructrados. Su desventaja es que tambien son mayores cajas negras y en la practica pueden ser tan o mas dificiles de implementar. Aun no queda claro cual es el mejor enfoque, si es que hay alguno, pueden leer una comparacion en [este link](http://www.mcgovern-fagg.org/amy/pubs/tr-06-001.pdf).

**Nosotros vamos a utilizar el el enfoque de funciones de valor**, pues da un equilibrio entre estructura y practicidad.

## Politcas optimas y $\epsilon$-greegy 

En la realidad, nuestros algoritmos no escogeran siempre la decision optima $\pi^*(S_t)$, pero si con una probabilidad $(1-\epsilon)$ donde $\epsilon$ ira decreciendo con el tiempo. El problema de escoger siempre la decision optima es que en la practica no sabemos a la perfeccion los parametros del MDP, y un agente que no explore, tiene el riesgo de llegar a conclusiones equivocadas prematuramente y nunca encontrar la politica optima. Este tipo de politicas se conocen como $\epsilon$-greedy.

## Implementacion en Julia de la funcion de valor Q


<span style="color: blue">**Advertencia!**  **Hasta nuevo aviso, supondremos que el conjunto de estados es discreto, pues aun no estamos en facultad de desarollar teoria para conjuntos de estados continuos.**.</span>

La forma mas tentadora de almacenar la informacion de $Q$ es utilizando una matriz o un vector. Pero usualmente es mucho mas conveniente usar *tablas hash*, que en Julia (y Python) son llamadas *diccionarios*. Una tabla hash es una estructura de datos que consiste de *claves* y *valores*, asignando a cada clave un valor. Es una estructura no ordenada, a diferencia de un vector. Estas son algunas ventajas

* Usualmente iniciamos los diccionarios vacios y agregamos claves la primera vez que ocurren. En problemas de reinforcement learning hay muchisimos estados que nunca son visitados y que resulta ineficiente o imposible almacenar un valor para cada estado (como en el ajedrez). Las tablas has permite solo asignar valores a las tuplas (estado, accion) observadas, mas similar a como aprendemos en la realidad.
* Con un diccionario los estados pueden ser claves aunque en si mismos sean arrays o objetos compuestos. Muchas otras estructuras necesitan de "nombres" en los estados, y por lo tanto solo pueden ser strings.

Para inicializar un diccionaro vacio en Julia usamos la funcion `Dict`

In [8]:
Q = Dict()

Dict{Any,Any} with 0 entries

Las parejas estado-accion se pueden modelar como tuplas, representadas en Julia con parentesis $(s,a)$. El siguiente ejemplo muestra como manipular $Q$ como un diccionario que tiene como claves las tuplas $(s,a)$. Se pueden usar brackets [ ] para acceder el valor de las claves del diccionario y para signar nuevos valores.

In [9]:
estado = "Some state"
accion = "Some action"
Q[(estado, accion)] = 1.
Q[(estado, accion)] 

1.0

Para acceder a las claves actualmente guardadas se puede usar el comando

In [10]:
keys(Q)

Base.KeyIterator for a Dict{Any,Any} with 1 entry. Keys:
  ("Some state","Some action")

Tambien podemos consultar si el diccionario tiene una clave

In [11]:
haskey(Q, (estado, accion))

true

Una funcion implementada comunmente es las tablas Hash es el metodo `get` que busca una clave y si no la encuentra regresa un valor *default*

In [12]:
default_value = 0.
get(Q, "I am not a key", default_value)

0.0

## Implementacion en Julia de una politica $\epsilon$-greedy en Julia

La implementacion de la politica de un agente puede ser implica o explicita
* Es *explicita* si dado un estado devuelve la distribucion de probabilidad de la politica
* Es *implicita* si dado un estado devuelve una observacion de la distribucion de la politica
Aunque matematicamente modelamos una politica como una distribucion de probabilidad para cada estado. En la practica es muchas veces conveniente modelarla implicitamente. Una de las ventajas de los metodos de funciones de valor es que funcionan muy bien con politicas implicitas, que no necesitan almacenar una gran cantidad de valores por cada combinacion de estado accion, si no que reciben directamente a la funcion de valor $Q$ como argumento una accion.

La politica debe consistir en los siguientes ingredientes:

> ### Ingredientes de una politica de tipo $\epsilon$-greedy:
> * **Inputs**
>    1. `state`: el estado actual en el cual se tiene que tomar una accion
>    2. `actions`: las acciones disponibles al estado `state`
>    3. `Q`: la mejor estimacion hasta el momento de la matriz `Q`, que se usara para tomar la decision greedy $$\mathrm{greedy}(\text{state}) = \mathrm{argmax}_{a\in\text{actions}} Q(\text{state}, a)$$
>    4. $\epsilon$: la probabilidad de tomar una decision completamente al azar y no greedy
> * **Output**
>    + La accion $\mathrm{policy}(\text{state})$ decidida por 
>        - aleatoriamente entre cualquier accion con probabilidad $\epsilon$ 
>        - aleatoriamente entre las acciones que maximicen $Q(\text{state}, \cdot)$ con probabilidad $1-\epsilon$.
        

In [13]:
"""
Returns an action according to an ϵ-greedy policy given an estimate of the action value function Q.
"""
function policy(state::Any, action_space::Any, Q::Dict, ϵ::Float64)
    random_number = rand()
    if random_number < ϵ
        return rand(action_space) #  totalmente aletorio
    else
        max_Q = maximum([get(Q, (state, a), 0.) for a in action_space]) # max_a Q(state, a)
        greedy_actions = [a for a in action_space if get(Q, (state, a), 0.) == max_Q] # argmax_a Q(state, a)
        return rand(greedy_actions) # aleatorio entre las mejores acciones solamente
    end
end

policy

**EJEMPLO DE USO**

In [14]:
Q = Dict()
ϵ = 0.25 # 25% de aleatoriedad total
state = "Some state"
action_space = 0:3
Q[(state, 0)] = Q[(state, 3)] = 1. # Las mejores acciones son 0 y 3
Q[(state, 1)] = Q[(state, 2)] = 0. # Las peores acciones son 1 y 2
for i in 1:30 # Imprime 30 simulaciones de acciones
    print(policy(state, action_space, Q, 0.1), " ")
end

3 0 3 3 3 3 0 3 3 3 0 0 3 0 3 3 3 3 0 0 0 3 0 0 3 3 3 0 0 0 

### <span style='color: red'> Ejercicio 2 </span>

1. Para el juego de MountainCar describe en una o dos oraciones lo que consideres es una politica optima
2. Para el juego de Pong describe en una o dos oraciones que consideres es una politica optima
3. Describe  en una o dos oraciones los potenciales peligros de tener una politica $\epsilon$-greedy en la que $\epsilon$ sea demasiado baja.
4. Resume en una o dos oraciones las diferencias entre las dos ecuaciones de Bellman.
5. Porque decimos que las ecuaciones de Bellman son recursivas?

## 3. Metodos tabulares de control optimo: Q-Learning y Sarsa

Como mencionamos antes, los metodos de control optimo de Reinforcement Learning que vamos a usar se basan en la estimacion de la funcion de valor $Q$ y para eso usan las Ecuaciones de Bellman y la Ley de los Grandes Numeros (LGN). Vamos a tratar de clarificar un poco mas la razon.

***Ley de los Grandes Numeros:* ** si se tiene una muestra de variables aleatorias independientes e identicas $X_1,...,X_n$ (con esperenza finita) entonces 
$$
\frac{1}{n}\sum_{i=1}^nX_i \stackrel{n\to\infty}{\to} E[X_i].
$$

Las exposiciones formales de los metodos de RL por funciones de valor y sus propiedades de convergencia puede ser muy extensa, pero la intuicion es increiblemente simple. Por ejemplo, para utilizar la LGN junto con la primera ecuacion de Bellman podemos hacer lo siguiente:

<blockquote>
<h3> Derivacion intuitiva de SARSA </h3>
<oi>
<li> Denotemos $Q_t$ la estimacion de $Q$ disponible al tiempo $t$. 
<li> Como 
$$E \left[ R_{t+1} + \gamma Q(S_{t+1}, A_{t+1}) \right] - Q(S_t, A_t) = 0$$
segun la primera ecuacion de Bellman, entonces las diferencia 
$$
\delta_t(S_t, A_t) = R_{t+1} + \gamma Q_t(S_{t+1}, A_{t+1})  - Q_t(S_t, A_t)
$$
nos da informacion de que tan bien o que tan mal esta la estimacion actual $Q_t(S_t, A_t)$.
<li>  Como las cantidades $\delta_t(S_t, A_t)$ son "en promedio" cero segun la LGN, entonces podemos dar pequeños pasas en la direccion $\delta_t(S_t, A_t)$ para mejorar nuestra estimacion. Conforme los estados $(S_t, A_t)$ sigan aparenciendo en el futuro, tendremos una buena estimacion de $Q(S_t, A_t)$. Matematicamente
$$
Q_t(S_t, A_t) + \alpha \delta_t \stackrel{t\to \infty}{ \to} Q(S_t, A_t).
$$
<li> La constante $0 < \alpha < 1$ se conoce como **velocidad de aprendizaje**. Valores tipicos son $\alpha \approx 0.025, 0.05, 0.1$ pero puede variar mucho de aplicacion en aplicacion.
<li>  Una interpretacion intuitiva de $\alpha$ es que define una nueva estimacion $Q_{t+1}(S_t, A_t)$, dandole un peso $(1-\alpha)$ a la estimacion actual $Q_t(S_t, A_t)$ y un peso $\alpha$ a la estimacion $R_{t+1} + \gamma Q_t(S_{t+1}, A_{t+1}) $.
</oi>
</blockquote>

La idea anterior se convierte en un metodo de control para encontrar una politica "optima" (ver el ejercicio debajo) llamado SARSA. El nombre SARSA se deriva de la regla de actualizacion de $Q$ que utiliza las variables $S_t, A_t, R_{t+1}, S_{t+1}, A_{t+1}$. La convergencia de este metodo (y tambien los casos en los que podria fallar) esta explicada por las teoria de *Aproximacion Estocastica*.


> ### SARSA 
>
> **Inputs**: 
  * Velocidad de aprendizaje $0 < \alpha < 1$
  * Tasa de descuento de la utilidad en el futuro $0 < \gamma \leq 1$ 
  * Una politica, tipicamente $\epsilon$-greedy.
>  
> **Outputs**
> * Una estimacion de la funcion de valor $Q$ bajo la politica $\epsilon$-greedy
>
> **Algoritmo**
> * **Inicializar** $Q$.
> * **Para cada episodio** con estado-accion inicial $(s_0, a_0)$:
>    + Para cada $t$ mientras el episodio no haya terminado:
>        - **Hacer una transicion** a un nuevo estado-accion $(s_{t+1}, a_{t+1})$ obteniendo un pago $r_{t+1}$.
>        - <span style='color: blue'>**Calcular la diferencia en utilidad segun la regla SARSA**</span>
>          $$ \delta_t = r_{t+1} + \gamma Q(s_{t+1}, a_{t+1}) - Q(s_t,a_t) $$
>        - **Actualizar el valor de $Q$**:
>            $$ Q(s,a) = Q(s,a) + \alpha\delta_t$$  
 
Cada vez que el algoritmo encuentre una pareja $(s,a)$ le debe asignar un valor por *default* a $Q(s,a)$. Lo tradicional es asignarle el valor $0$. Las propiedades de convergencia estocastico hacen que no importe este valor inicial.

SARSA es un metodo **tabular** porque Q se calcular para cada posible tupla estado-accion. En la proxima tarea veran metodos aproximados que reducen el espacio de estados a un numero selecto de *features*.


### <span style='color: red'> Ejercicio 3 </span>

1. Con SARSA se obtiene un estimacion de la funcion de valor $Q$ de la politica optima? Si no se obtiene la politica optima, que politica se obtiene? Es necesariamente malo una politica suboptima durante el proceso de aprendizaje?
2. **(Q-LEARNING)** El famoso algoritmo Q-Learning (Watkins 1989) es identico al de SARSA salvo en una linea, reemplazamos la regla SARSA para $\delta_t$ por
$$ \delta_t = r_{t+1} + \gamma \max_a Q_{\pi^*}(s_{t+1}, a) - Q(s_t,a_t) $$
Imitando la derivacion intuitiva de SARSA, explica intuitivamente el algoritmo Q-Learning usando la ecuacion de optimalidad de Bellman.
3. SARSA se clasifica como un metodo de control *on-policy* y Q-Learning como *off-policy*. La razon es que no estiman la misma Q: uno de los metodos estima la optima y el otro no, incluso si ambos se implementan usando la misma politica $\epsilon$-greedy. Investiga en maximo un parrafo mas sobre la diferencia entre SARSA y Q-Learning (o entre control *on-policy* y *off-policy*) y contesta: Es mejor un metodo off-policy que un metodo on-policy?
4. Hay diferencia entre SARSA y Q-Learning conforme $\epsilon \to 0$?



## 4. Let's play

Vamos a implementar el algoritmo SARSA para resolver el problema del coche en la colina. Cual es el primer problema que enfrentaremos? Preprocesar los estados.

### Preprocesamiento de estados

Este es el paso que en practicamente ninguna aplicacion de RL podemos escapar. Ademas, los distintos retos dependen del tipo de problema. Algunos puntos comunes que puede surgir son
* **El espacio de estados es demasiado grande?**. Un ejemplo tipico es querer aprender juegos de Atari a traves de observar pixeles, pues muchos pixeles pueden hacer el conjunto de estados inmanejable. En la proxima tarea veremos metodos aproximados que ayudan a resolver el problema de un conjunto de estados muy grande creando features. En metodos tabulares hay que reducir el tamano del conjunto de estados, por ejemplo, en el caso de imagenes de Atari, comprimiendo el tamano de las imagenes y solo guardando la informacion mas importante de los pixeles.
* **El espacio de estados es continuo?**. Si vamos a usar un metodo tabular como SARSA o Q-Learning sera necesario discretizar en un numero finito de valores. Debajo hacemos una demostracion. Nuevamente, los metodos aproximados pueden evitar la necesidad de discretizar.
* **El espacio de estados es markoviano?**. Checar esta condicion es vital y usualmente ignorada. El funcionamiento correcto de los metodos depende del supuesto de Markov, i.e., que el futuro rendimiento dependa unicamente del estado actual y de la decision tomada. De manera practica podemos preguntar: conocer el ultimo estado es suficiente para tomar una decision optima?

La siguiente funcion hace una discretizacion simple de un conjunto de estados de dimension $k$, seleccionando el valor minomo, maximo y el numero de puntos en cada dimension, asumiendo una discretizacion uniforme. Se pueden hacer funciones de discretizacion mucho mas sofisticadas.

In [15]:
"""
For each index `k` of the array `state`, it creates an even grid of length npoints[`k`] 
going from `lower[k]` to `upper[k]`, and returns the closest value of `state[k]` on that grid. 
All the arguments must be of the same length.
"""
function discretize(state::Array, lower::Array, upper::Array, npoints::Array) 
    discretized = Array{Float32}(length(state)) # preallocating is faster (but not necessary)
    for k in 1:length(state)
        grid = linspace(lower[k], upper[k], npoints[k]) # even space
        discretized[k] = grid[indmin(abs(grid - state[k]))] # closest point
    end
    return discretized
end
;

**EJEMPLO DE USO**

In [16]:
function preprocess(state::Array)
    lower = [-1.2, -0.07]
    upper = [0.6, 0.07]
    npoints = [16, 8]
    discretize(state, lower, upper, npoints)
end

mountain_car = gym.make("MountainCar-v0")
state = mountain_car[:reset]()
println(state, " becomes ", preprocess(state))

[-

[2017-05-10 22:24:56,364] Making new env: MountainCar-v0


0.460062,0.0] becomes Float32[-0.48,-0.01]


### Funcion de entrenamiento

Necesitamos una funcion que reciba una estimacion actual de la tabla $Q$ y que devuelva una estimacion mejorada. Notemos que la funcion depende del preprocesamiento que definimos anteriormente.

**APRENDIENDO UN EPISODIO**

In [17]:
"""
The following function plays a gym_episode and learns using SARSA. It receives a current estimate of Q and 
returns an improved estimate Q, the episode reward and the number of steps.
"""
function learn_episode_with_sarsa(
        gym_env, # an openai gym environment
        current_Q = Dict() # current estimation of Q
        ; # the rest are keyword argument, not positional arguments
        preprocess_state = x -> x, # state preprocessing, identity by default
        learning_rate = 0.1,
        discount_rate = 0.999,
        exploration_rate = 0.1,
        render = false, # show game visualization (slower when rendering...)
    )
    # Rename variables for formula simplicity following Sutton
    ϵ, α, γ = exploration_rate, learning_rate, discount_rate
    action_space = 0:(gym_env[:action_space][:n] - 1)
    Q = deepcopy(current_Q) # Q will store the improved Q estimate
    # Reset environment
    state = gym_env[:reset]() # regresa el ambiente al comienzo de un episodio
    state = preprocess_state(state)
    render && gym_env[:render]() # si render es true, crea la visualizacion del estado del juego
    # Get initial action
    action = policy(state, action_space, Q, ϵ) # selecionar accion con politica
    !haskey(Q, (state, action)) && (Q[(state, action)] = 0.) # add key if missing
    # Episode iteration information
    done = false # sera true cuando termine el episodio
    ep_reward = 0. # iniciar pago total en el episodio
    ep_steps = 0 # iniciar pasos totales en el episodio
    while !done 
        # Simulate transition
        new_state, reward, done, info = gym_env[:step](action) # simula una transicion
        new_state = preprocess_state(new_state)
        new_action = policy(new_state, action_space, Q, ϵ) # selecionar accion con politica
        !haskey(Q, (new_state, new_action)) && (Q[(new_state, new_action)] = 0.) # add key if missing
        # Compute δ
        δ = reward + γ*Q[(new_state, new_action)]  - Q[(state, action)] 
        # Update Q 
        Q[(state, action)] = Q[(state, action)] + α*δ
        # Change new states to old states for next step
        state = new_state
        action = new_action
        # Save step information and render if needed
        render && gym_env[:render]()
        ep_reward += reward # agrega el pago de la transicion al pago del ep.
        ep_steps += 1 # agrega un paso al total de pasos en el episodio
    end
    render && gym_env[:render](close = true)
    return Q, ep_reward, ep_steps
end

**EJEMPLO DE USO**

In [18]:
Q = Dict()
Q, reward, steps = learn_episode_with_sarsa(mountain_car, Q, preprocess_state = preprocess, render = true)

(Dict{Any,Any}(Pair{Any,Any}((Float32[-0.6,0.01],0),-1.45514),Pair{Any,Any}((Float32[-0.36,-0.01],1),-0.207947),Pair{Any,Any}((Float32[-0.48,0.01],1),-1.80265),Pair{Any,Any}((Float32[-0.6,-0.01],2),-1.20812),Pair{Any,Any}((Float32[-0.6,-0.01],0),-1.257),Pair{Any,Any}((Float32[-0.48,-0.01],1),-1.62641),Pair{Any,Any}((Float32[-0.6,-0.01],1),-1.37677),Pair{Any,Any}((Float32[-0.48,0.01],0),-1.7589),Pair{Any,Any}((Float32[-0.48,-0.01],0),-1.54719),Pair{Any,Any}((Float32[-0.36,0.01],0),-0.1)…),-200.0,200)

**APRENDIENDO DE MUCHOS EPISODIOS Y REDUCIENDO TASA DE EXPLORACION**

In [19]:
"""
This function will train a gym environment for many episodes using sarsa and a preprocess
function given by the user, defaults to identity.
"""
function learn_episodes_with_sarsa(
        n_episodes, # number of episodes
        gym_env, # an openai gym environment
        current_Q = Dict() # current estimation of Q
        ; # the rest are keyword argument, not positional arguments
        preprocess_state = x -> x, # state preprocessing, identity by default
        learning_rate = 0.1,
        discount_rate = 0.999,
        exploration_rate = 0.1,
        exploration_decay =  1 - log(2) / n_episodes, # se reducira a la mitad al final del episodio
        render = false, # show game visualization (slower when rendering...)
        verbose = true
    )
    Q = deepcopy(current_Q)
    # Iterate episodes
    reward_history = []
    steps_history = []
    for i in 1:n_episodes
        # learn from episode
        Q, reward, steps = learn_episode_with_sarsa(gym_env, Q, 
                                                    learning_rate = learning_rate, 
                                                    discount_rate = discount_rate, 
                                                    exploration_rate = exploration_rate, 
                                                    render = render,
                                                    preprocess_state = preprocess_state)
        verbose && println("Episodio ", i, " con pago ", reward)
        # reduce exploration_rate
        exploration_rate *= exploration_decay
        # save episode information
        push!(reward_history, reward)
        push!(steps_history, steps)
    end
    return Q, reward_history, steps_history
end

In [20]:
Q, reward_history, steps_history = learn_episodes_with_sarsa(
    5000, mountain_car, Dict(), 
    preprocess_state = preprocess,
    learning_rate = 0.05,
    exploration_rate = 0.25,
    exploration_decay = 1 - log(25) / 5000, # will reduce to 0.01 at the end of training
    render = true # be patient, enjoy watching, switch to false for speed
)

Episodio 1 con pago -200.0
Episodio 2 con pago -200.0
Episodio 3 con pago -200.0
Episodio 4 con pago -200.0
Episodio 5 con pago -200.0
Episodio 6 con pago -200.0
Episodio 7 con pago -200.0
Episodio 8 con pago -200.0
Episodio 9 con pago -200.0
Episodio 10 con pago -200.0
Episodio 11 con pago -200.0
Episodio 12 con pago -200.0
Episodio 13 con pago -200.0
Episodio 14 con pago -200.0
Episodio 15 con pago -200.0
Episodio 16 con pago -200.0
Episodio 17 con pago -200.0
Episodio 18 con pago -200.0
Episodio 19 con pago -200.0
Episodio 20 con pago -200.0
Episodio 21 con pago -200.0
Episodio 22 con pago -200.0
Episodio 23 con pago -200.0
Episodio 24 con pago -200.0
Episodio 25 con pago -200.0
Episodio 26 con pago -200.0
Episodio 27 con pago -200.0
Episodio 28 con pago -200.0
Episodio 29 con pago -200.0
Episodio 30 con pago -200.0
Episodio 31 con pago -200.0
Episodio 32 con pago -200.0
Episodio 33 con pago -200.0
Episodio 34 con pago -200.0
Episodio 35 con pago -200.0
Episodio 36 con pago -200.0
E

Episodio 288 con pago -200.0
Episodio 289 con pago -200.0
Episodio 290 con pago -200.0
Episodio 291 con pago -200.0
Episodio 292 con pago -200.0
Episodio 293 con pago -200.0
Episodio 294 con pago -200.0
Episodio 295 con pago -200.0
Episodio 296 con pago -200.0
Episodio 297 con pago -200.0
Episodio 298 con pago -200.0
Episodio 299 con pago -200.0
Episodio 300 con pago -200.0
Episodio 301 con pago -200.0
Episodio 302 con pago -200.0
Episodio 303 con pago -200.0
Episodio 304 con pago -200.0
Episodio 305 con pago -200.0
Episodio 306 con pago -200.0
Episodio 307 con pago -200.0
Episodio 308 con pago -200.0
Episodio 309 con pago -200.0
Episodio 310 con pago -200.0
Episodio 311 con pago -200.0
Episodio 312 con pago -200.0
Episodio 313 con pago -200.0
Episodio 314 con pago -200.0
Episodio 315 con pago -200.0
Episodio 316 con pago -200.0
Episodio 317 con pago -200.0
Episodio 318 con pago -200.0
Episodio 319 con pago -200.0
Episodio 320 con pago -200.0
Episodio 321 con pago -200.0
Episodio 322 c

Episodio 571 con pago -200.0
Episodio 572 con pago -200.0
Episodio 573 con pago -200.0
Episodio 574 con pago -200.0
Episodio 575 con pago -200.0
Episodio 576 con pago -200.0
Episodio 577 con pago -200.0
Episodio 578 con pago -200.0
Episodio 579 con pago -200.0
Episodio 580 con pago -200.0
Episodio 581 con pago -200.0
Episodio 582 con pago -200.0
Episodio 583 con pago -200.0
Episodio 584 con pago -200.0
Episodio 585 con pago -200.0
Episodio 586 con pago -200.0
Episodio 587 con pago -200.0
Episodio 588 con pago -200.0
Episodio 589 con pago -200.0
Episodio 590 con pago -200.0
Episodio 591 con pago -200.0
Episodio 592 con pago -200.0
Episodio 593 con pago -200.0
Episodio 594 con pago -200.0
Episodio 595 con pago -200.0
Episodio 596 con pago -200.0
Episodio 597 con pago -200.0
Episodio 598 con pago -200.0
Episodio 599 con pago -200.0
Episodio 600 con pago -200.0
Episodio 601 con pago -200.0
Episodio 602 con pago -200.0
Episodio 603 con pago -200.0
Episodio 604 con pago -200.0
Episodio 605 c

Episodio 854 con pago -200.0
Episodio 855 con pago -180.0
Episodio 856 con pago -190.0
Episodio 857 con pago -200.0
Episodio 858 con pago -200.0
Episodio 859 con pago -200.0
Episodio 860 con pago -200.0
Episodio 861 con pago -200.0
Episodio 862 con pago -200.0
Episodio 863 con pago -200.0
Episodio 864 con pago -200.0
Episodio 865 con pago -200.0
Episodio 866 con pago -164.0
Episodio 867 con pago -200.0
Episodio 868 con pago -164.0
Episodio 869 con pago -152.0
Episodio 870 con pago -200.0
Episodio 871 con pago -200.0
Episodio 872 con pago -200.0
Episodio 873 con pago -168.0
Episodio 874 con pago -200.0
Episodio 875 con pago -200.0
Episodio 876 con pago -165.0
Episodio 877 con pago -200.0
Episodio 878 con pago -200.0
Episodio 879 con pago -200.0
Episodio 880 con pago -200.0
Episodio 881 con pago -200.0
Episodio 882 con pago -164.0
Episodio 883 con pago -200.0
Episodio 884 con pago -194.0
Episodio 885 con pago -200.0
Episodio 886 con pago -200.0
Episodio 887 con pago -200.0
Episodio 888 c

Episodio 1132 con pago -200.0
Episodio 1133 con pago -200.0
Episodio 1134 con pago -200.0
Episodio 1135 con pago -200.0
Episodio 1136 con pago -200.0
Episodio 1137 con pago -200.0
Episodio 1138 con pago -200.0
Episodio 1139 con pago -200.0
Episodio 1140 con pago -200.0
Episodio 1141 con pago -200.0
Episodio 1142 con pago -200.0
Episodio 1143 con pago -200.0
Episodio 1144 con pago -200.0
Episodio 1145 con pago -200.0
Episodio 1146 con pago -200.0
Episodio 1147 con pago -200.0
Episodio 1148 con pago -200.0
Episodio 1149 con pago -200.0
Episodio 1150 con pago -200.0
Episodio 1151 con pago -200.0
Episodio 1152 con pago -200.0
Episodio 1153 con pago -200.0
Episodio 1154 con pago -200.0
Episodio 1155 con pago -200.0
Episodio 1156 con pago -200.0
Episodio 1157 con pago -200.0
Episodio 1158 con pago -200.0
Episodio 1159 con pago -173.0
Episodio 1160 con pago -200.0
Episodio 1161 con pago -200.0
Episodio 1162 con pago -200.0
Episodio 1163 con pago -200.0
Episodio 1164 con pago -148.0
Episodio 1

Episodio 1406 con pago -200.0
Episodio 1407 con pago -200.0
Episodio 1408 con pago -200.0
Episodio 1409 con pago -200.0
Episodio 1410 con pago -200.0
Episodio 1411 con pago -200.0
Episodio 1412 con pago -200.0
Episodio 1413 con pago -200.0
Episodio 1414 con pago -200.0
Episodio 1415 con pago -200.0
Episodio 1416 con pago -200.0
Episodio 1417 con pago -200.0
Episodio 1418 con pago -200.0
Episodio 1419 con pago -200.0
Episodio 1420 con pago -200.0
Episodio 1421 con pago -200.0
Episodio 1422 con pago -200.0
Episodio 1423 con pago -151.0
Episodio 1424 con pago -200.0
Episodio 1425 con pago -200.0
Episodio 1426 con pago -200.0
Episodio 1427 con pago -200.0
Episodio 1428 con pago -200.0
Episodio 1429 con pago -200.0
Episodio 1430 con pago -200.0
Episodio 1431 con pago -200.0
Episodio 1432 con pago -200.0
Episodio 1433 con pago -200.0
Episodio 1434 con pago -200.0
Episodio 1435 con pago -200.0
Episodio 1436 con pago -200.0
Episodio 1437 con pago -200.0
Episodio 1438 con pago -200.0
Episodio 1

Episodio 1953 con pago -200.0
Episodio 1954 con pago -200.0
Episodio 1955 con pago -163.0
Episodio 1956 con pago -200.0
Episodio 1957 con pago -200.0
Episodio 1958 con pago -200.0
Episodio 1959 con pago -200.0
Episodio 1960 con pago -200.0
Episodio 1961 con pago -159.0
Episodio 1962 con pago -200.0
Episodio 1963 con pago -200.0
Episodio 1964 con pago -162.0
Episodio 1965 con pago -166.0
Episodio 1966 con pago -200.0
Episodio 1967 con pago -200.0
Episodio 1968 con pago -200.0
Episodio 1969 con pago -200.0
Episodio 1970 con pago -156.0
Episodio 1971 con pago -162.0
Episodio 1972 con pago -169.0
Episodio 1973 con pago -200.0
Episodio 1974 con pago -200.0
Episodio 1975 con pago -200.0
Episodio 1976 con pago -156.0
Episodio 1977 con pago -200.0
Episodio 1978 con pago -156.0
Episodio 1979 con pago -200.0
Episodio 1980 con pago -158.0
Episodio 1981 con pago -200.0
Episodio 1982 con pago -200.0
Episodio 1983 con pago -200.0
Episodio 1984 con pago -200.0
Episodio 1985 con pago -200.0
Episodio 1

Episodio 2500 con pago -129.0
Episodio 2501 con pago -127.0
Episodio 2502 con pago -200.0
Episodio 2503 con pago -152.0
Episodio 2504 con pago -200.0
Episodio 2505 con pago -140.0
Episodio 2506 con pago -200.0
Episodio 2507 con pago -200.0
Episodio 2508 con pago -132.0
Episodio 2509 con pago -200.0
Episodio 2510 con pago -171.0
Episodio 2511 con pago -200.0
Episodio 2512 con pago -178.0
Episodio 2513 con pago -173.0
Episodio 2514 con pago -168.0
Episodio 2515 con pago -175.0
Episodio 2516 con pago -171.0
Episodio 2517 con pago -166.0
Episodio 2518 con pago -200.0
Episodio 2519 con pago -200.0
Episodio 2520 con pago -155.0
Episodio 2521 con pago -200.0
Episodio 2522 con pago -154.0
Episodio 2523 con pago -149.0
Episodio 2524 con pago -200.0
Episodio 2525 con pago -200.0
Episodio 2526 con pago -153.0
Episodio 2527 con pago -157.0
Episodio 2528 con pago -200.0
Episodio 2529 con pago -200.0
Episodio 2530 con pago -200.0
Episodio 2531 con pago -200.0
Episodio 2532 con pago -200.0
Episodio 2

Episodio 2774 con pago -174.0
Episodio 2775 con pago -200.0
Episodio 2776 con pago -194.0
Episodio 2777 con pago -155.0
Episodio 2778 con pago -179.0
Episodio 2779 con pago -170.0
Episodio 2780 con pago -173.0
Episodio 2781 con pago -189.0
Episodio 2782 con pago -147.0
Episodio 2783 con pago -163.0
Episodio 2784 con pago -200.0
Episodio 2785 con pago -173.0
Episodio 2786 con pago -152.0
Episodio 2787 con pago -149.0
Episodio 2788 con pago -143.0
Episodio 2789 con pago -196.0
Episodio 2790 con pago -187.0
Episodio 2791 con pago -147.0
Episodio 2792 con pago -143.0
Episodio 2793 con pago -199.0
Episodio 2794 con pago -199.0
Episodio 2795 con pago -200.0
Episodio 2796 con pago -155.0
Episodio 2797 con pago -200.0
Episodio 2798 con pago -200.0
Episodio 2799 con pago -200.0
Episodio 2800 con pago -200.0
Episodio 2801 con pago -200.0
Episodio 2802 con pago -200.0
Episodio 2803 con pago -200.0
Episodio 2804 con pago -200.0
Episodio 2805 con pago -200.0
Episodio 2806 con pago -200.0
Episodio 2

Episodio 3320 con pago -200.0
Episodio 3321 con pago -159.0
Episodio 3322 con pago -162.0
Episodio 3323 con pago -146.0
Episodio 3324 con pago -152.0
Episodio 3325 con pago -186.0
Episodio 3326 con pago -171.0
Episodio 3327 con pago -163.0
Episodio 3328 con pago -157.0
Episodio 3329 con pago -154.0
Episodio 3330 con pago -166.0
Episodio 3331 con pago -200.0
Episodio 3332 con pago -158.0
Episodio 3333 con pago -159.0
Episodio 3334 con pago -156.0
Episodio 3335 con pago -164.0
Episodio 3336 con pago -157.0
Episodio 3337 con pago -161.0
Episodio 3338 con pago -162.0
Episodio 3339 con pago -166.0
Episodio 3340 con pago -159.0
Episodio 3341 con pago -153.0
Episodio 3342 con pago -163.0
Episodio 3343 con pago -158.0
Episodio 3344 con pago -167.0
Episodio 3345 con pago -172.0
Episodio 3346 con pago -167.0
Episodio 3347 con pago -179.0
Episodio 3348 con pago -166.0
Episodio 3349 con pago -174.0
Episodio 3350 con pago -167.0
Episodio 3351 con pago -156.0
Episodio 3352 con pago -158.0
Episodio 3

Episodio 3594 con pago -200.0
Episodio 3595 con pago -169.0
Episodio 3596 con pago -152.0
Episodio 3597 con pago -200.0
Episodio 3598 con pago -200.0
Episodio 3599 con pago -197.0
Episodio 3600 con pago -200.0
Episodio 3601 con pago -200.0
Episodio 3602 con pago -200.0
Episodio 3603 con pago -200.0
Episodio 3604 con pago -200.0
Episodio 3605 con pago -200.0
Episodio 3606 con pago -189.0
Episodio 3607 con pago -200.0
Episodio 3608 con pago -200.0
Episodio 3609 con pago -143.0
Episodio 3610 con pago -200.0
Episodio 3611 con pago -131.0
Episodio 3612 con pago -139.0
Episodio 3613 con pago -152.0
Episodio 3614 con pago -169.0
Episodio 3615 con pago -153.0
Episodio 3616 con pago -153.0
Episodio 3617 con pago -169.0
Episodio 3618 con pago -200.0
Episodio 3619 con pago -193.0
Episodio 3620 con pago -200.0
Episodio 3621 con pago -181.0
Episodio 3622 con pago -156.0
Episodio 3623 con pago -200.0
Episodio 3624 con pago -200.0
Episodio 3625 con pago -200.0
Episodio 3626 con pago -200.0
Episodio 3

Episodio 4140 con pago -200.0
Episodio 4141 con pago -200.0
Episodio 4142 con pago -200.0
Episodio 4143 con pago -200.0
Episodio 4144 con pago -200.0
Episodio 4145 con pago -200.0
Episodio 4146 con pago -161.0
Episodio 4147 con pago -200.0
Episodio 4148 con pago -159.0
Episodio 4149 con pago -200.0
Episodio 4150 con pago -200.0
Episodio 4151 con pago -200.0
Episodio 4152 con pago -200.0
Episodio 4153 con pago -200.0
Episodio 4154 con pago -200.0
Episodio 4155 con pago -200.0
Episodio 4156 con pago -166.0
Episodio 4157 con pago -177.0
Episodio 4158 con pago -200.0
Episodio 4159 con pago -198.0
Episodio 4160 con pago -168.0
Episodio 4161 con pago -175.0
Episodio 4162 con pago -173.0
Episodio 4163 con pago -168.0
Episodio 4164 con pago -200.0
Episodio 4165 con pago -200.0
Episodio 4166 con pago -168.0
Episodio 4167 con pago -166.0
Episodio 4168 con pago -187.0
Episodio 4169 con pago -168.0
Episodio 4170 con pago -200.0
Episodio 4171 con pago -171.0
Episodio 4172 con pago -175.0
Episodio 4

Episodio 4414 con pago -169.0
Episodio 4415 con pago -187.0
Episodio 4416 con pago -191.0
Episodio 4417 con pago -149.0
Episodio 4418 con pago -140.0
Episodio 4419 con pago -200.0
Episodio 4420 con pago -141.0
Episodio 4421 con pago -200.0
Episodio 4422 con pago -117.0
Episodio 4423 con pago -199.0
Episodio 4424 con pago -140.0
Episodio 4425 con pago -138.0
Episodio 4426 con pago -138.0
Episodio 4427 con pago -200.0
Episodio 4428 con pago -148.0
Episodio 4429 con pago -200.0
Episodio 4430 con pago -200.0
Episodio 4431 con pago -137.0
Episodio 4432 con pago -137.0
Episodio 4433 con pago -136.0
Episodio 4434 con pago -200.0
Episodio 4435 con pago -200.0
Episodio 4436 con pago -200.0
Episodio 4437 con pago -142.0
Episodio 4438 con pago -141.0
Episodio 4439 con pago -138.0
Episodio 4440 con pago -187.0
Episodio 4441 con pago -118.0
Episodio 4442 con pago -138.0
Episodio 4443 con pago -137.0
Episodio 4444 con pago -127.0
Episodio 4445 con pago -139.0
Episodio 4446 con pago -112.0
Episodio 4

Episodio 4688 con pago -135.0
Episodio 4689 con pago -149.0
Episodio 4690 con pago -143.0
Episodio 4691 con pago -145.0
Episodio 4692 con pago -151.0
Episodio 4693 con pago -152.0
Episodio 4694 con pago -152.0
Episodio 4695 con pago -200.0
Episodio 4696 con pago -153.0
Episodio 4697 con pago -149.0
Episodio 4698 con pago -147.0
Episodio 4699 con pago -200.0
Episodio 4700 con pago -153.0
Episodio 4701 con pago -154.0
Episodio 4702 con pago -146.0
Episodio 4703 con pago -149.0
Episodio 4704 con pago -146.0
Episodio 4705 con pago -152.0
Episodio 4706 con pago -140.0
Episodio 4707 con pago -147.0
Episodio 4708 con pago -149.0
Episodio 4709 con pago -136.0
Episodio 4710 con pago -143.0
Episodio 4711 con pago -144.0
Episodio 4712 con pago -200.0
Episodio 4713 con pago -136.0
Episodio 4714 con pago -190.0
Episodio 4715 con pago -140.0
Episodio 4716 con pago -136.0
Episodio 4717 con pago -138.0
Episodio 4718 con pago -156.0
Episodio 4719 con pago -200.0
Episodio 4720 con pago -200.0
Episodio 4

(Dict{Any,Any}(Pair{Any,Any}((Float32[-0.24,-0.05],0),-99.9185),Pair{Any,Any}((Float32[-0.48,-0.07],0),-55.1594),Pair{Any,Any}((Float32[0.36,0.05],2),-13.1868),Pair{Any,Any}((Float32[0.0,0.03],2),-86.5017),Pair{Any,Any}((Float32[-0.96,0.03],0),-101.32),Pair{Any,Any}((Float32[-1.08,-0.01],0),-103.93),Pair{Any,Any}((Float32[-0.96,0.03],2),-101.327),Pair{Any,Any}((Float32[-0.36,0.05],1),-92.3451),Pair{Any,Any}((Float32[0.0,0.05],2),-77.316),Pair{Any,Any}((Float32[-0.24,-0.03],1),-115.862)…),Any[-200.0,-200.0,-200.0,-200.0,-200.0,-200.0,-200.0,-200.0,-200.0,-200.0  …  -200.0,-200.0,-200.0,-200.0,-200.0,-200.0,-200.0,-200.0,-200.0,-200.0],Any[200,200,200,200,200,200,200,200,200,200  …  200,200,200,200,200,200,200,200,200,200])

**EJEMPLO DE USO**

In [21]:
using Plots; gr()
plot_every = 100
roll_mean = []
for i in 1:50
   push!(roll_mean, mean(reward_history[(plot_every*(i-1)+1):(plot_every*i)])) 
end
plot(roll_mean, title = "Retorno promedio en batches de 100 episodios")

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 10 
 
 
 20 
 
 
 30 
 
 
 40 
 
 
 50 
 
 
 -200 
 
 
 -190 
 
 
 -180 
 
 
 -170 
 
 
 -160 
 
 
 Retorno promedio en batches de 100 episodios 
 
<polyline clip-path="url(#clip02)" style="stroke:#0099ff; stroke-width:0.8; stroke-opacity:1; fill:none" points="
 39.4926,384.952 50.8511,384.952 62.2097,384.274 73.5683,382.014 84.9269,381.261 96.2855,379.98 107.644,382.918 119.003,349.171 130.361,346.234 141.72,367.551 
 153.078,370.112 164.437,372.448 175.796,348.644 187.154,363.258 198.513,352.938 209.871,349.021 221.23,282.507 232.588,197.387 243.947,280.096 255.306,273.844 
 266.664,254.937 278.023,198.517 289.381,288.684 300.74,166.428 312.098,77.1651 323.457,313.768 334.816,187.143 346.174,248.082 357.533,297.12 368.891,185.862 
 380.25,292.149 391.609,167.407 402.967,168.763 414.326,171.098 425.684,168.612 437.043,313.165 448.401,153.17 459.76,257.122 471.119,126.128 482.477,294.333 
 493.836,207.632 505.194,241.981 516.553,155.882 527.912,148.877 539.27,71.9675 550.629,66.1673 561.987,167.181 573.346,23.3815 584.704,218.554 596.063,324.991 
 
 "/>
 
 
 
 
 y1

### <span style='color: red'> Ejercicio 4 </span>

1. Replica el codigo anterior de entrenamiento en 5000 episodios haciendo los cambios necesarios para aprender con Q-Learning en vez de SARSA.